### ISSUE TITLE
Please keep bumping the latest release version

### ISSUE THREAD

**Issue by DanVanAtta** [triplea-game/triplea#2126]
For those with write access, when merging a backward compatible change, as a matter of practice please remember to come back and mark the latest release as 'stable'.

**Comment by ssoloff** [comment]
@DanVanAtta Does this include only functional changes or _all_ backwards-compatible changes?  For example, if someone merges non-functional Checkstyle warning fixes, should that trigger a new stable release?

**Comment by DanVanAtta** [comment]
Good question. I think it would be better to update the latest version after all compatible updates are merged, functional or not. In practice, this probably would be done once at the end of the day. Release notes are useful and interesting for a number of users, it would be great to see more automation there.

It could be worth considering flipping the value on the travis prerelease build flag while we are accepting compatible only PRs: https://github.com/triplea-game/triplea/blob/master/.travis.yml#L40

**Comment by ssoloff** [comment]
> It could be worth considering flipping the value on the travis prerelease build flag while we are accepting compatible only PRs

:+1: I had assumed that we were marking Travis deployments as pre-releases by default because there needed to be some sort of manual acceptance testing performed before declaring an official release.  That doesn't appear to be the case.

There's also the possibility of using a gitflow-like style where we have both `develop` and `master` branches.  `develop` becomes the mainline branch for merging PRs, and Travis is configured to use `prerelease: true` when deploying from this branch.  `master` becomes the stable branch, and Travis is configured to use `prerelease: false` when deploying from this branch.  Then a simple merge from `develop` to `master` is what triggers a new official release.

> it would be great to see more automation there

Yeah, the biggest problem is probably keeping the release notes updated.

If we don't want to automate that right now, one solution is to require PR authors themselves to incrementally update the release notes (e.g. see [here](http://keepachangelog.com/en/1.0.0/)).  What makes that a bit inefficient at this time is that the release notes are kept in a separate repo, so it would require opening a second PR.

If we want to pursue automation, it may require using a commit message style that is amenable to automatic generation from the git log (e.g. see [here](https://desmart.com/blog/how-to-generate-your-project-s-changelog-from-commit-messages)).  That might also require a review of how we merge PRs containing more than one commit.  That is, would we have to specifically create a merge commit or squash before merging to ensure the correct commit message contains the changelog tag?

**Comment by DanVanAtta** [comment]
>  manual acceptance testing performed before declaring an official release

This is actually a bit nuanced. If there is a lot of time between releases, then we need to do full regression testing, which is largely manual. If the release delta is small, and since we manually test each branch before merge as part of PR acceptance, then we would have hit the testing bar after merge.

> There's also the possibility of using a gitflow-like style where we have both develop and master

There were some early discussions on which work flow to use. We settled on a single branch. I am still pretty opposed to the gitflow style. In our case there are some hurdles:
- if we have a develop branch, how do we create and push artifacts from it? If we don't do that, then is there any way we can run the release process frequently to know it will work? Without artifacts, it seems like it would be difficult to test
- What is the benefit of having the second branch? Today the release is a check of a box, we are considering to remove that. Merging develop into master introduces the risk of extra changes from merges (which granted ideally is an empty set, but since sometimes you have to get somethign on to master and the develop stuff conflicts too badly, in practice this introduces merge conflicts).

> Yeah, the biggest problem is probably keeping the release notes updated. 
> If we don't want to automate that right now, one solution is to require PR authors themselves to incrementally update the release notes (e.g. see here). What makes that a bit inefficient at this time is that the release notes are kept in a separate repo, so it would require opening a second PR.

I agree, though I suspect if we do not automate it, we will not always follow the process. Particularly when merge happens offline compared to when the author is around. An automated process would be okay since we could go back and hand edit the file to keep it reading well. 

> that might also require a review of how we merge PRs containing more than one commit.

Indeed, if we did so we would probably always want to squash merge, the title could be programmatically added perhaps to the release notes.

**Comment by RoiEXLab** [comment]
@DanVanAtta I'm not a huge fan of releasing every build as a new version.
It happens way too often that small one-line-fixable issues slip in, an example for this is #2133
Automation for automatic releases is good, auto-generating changelogs based on commits (or PR titles if that is possible) seems like a good idea to me.

**Comment by ssoloff** [comment]
@DanVanAtta Thanks for the historical information.

> if we have a develop branch, how do we create and push artifacts from it?

I understand your point about the two-branch solution not being ideal for TripleA.  Just wanted to mention that you can configure Travis to deploy artifacts from _multiple_ branches.  I'm actually doing this for another project where artifacts from builds of the `develop` branch get deployed to a Heroku staging environment, while artifacts from builds of the `master` branch get deployed to a Heroku production environment.

**Comment by DanVanAtta** [comment]
@RoiEXLab @ssoloff thanks for the discussion, this is actually a bit of a deep topic.

### Release frequency and not catching bugs are not necessarily related, slow releases do not mean we will catch important bugs (in recent past experience, regression testing has still missed  pretty important regressions)


> It happens way too often that small one-line-fixable issues slip in, an example for this is #2133

An interesting observation. I don't think release frequency necessarily would make that problem worse. For #2133 we discovered the bug after release. By releasing less frequently it would seem that we would delay that time to discovery, making the problem a bit worse even.

There is a deeper issue though having to do with the brittleness of the code and the lack of tests, and the difficulty we have in getting reasonable tests in place. With slow releases, we basically do a large regression test before the release and hope to catch all such issues. Past experience has been:
- sometimes the regression testing or partially skipped, we've missed quite a bit of stuff
- we can still see new issues get introduced by new changes between regression testing and release
- regression testing tends to be a 'stop-the-world' and let the PR queue back up for a month. Merge conflicts could get bad, in practice development takes a pause regardless if there are other new contributors or small fixes to be applied. We could look to solutions like a 'develop/master' branch to avoid 'pause-the-world', but we would still be talking a considerable amount of new overhead to something that is a non-problem if we simply release frequently.

We saw all of the above problems between the 1.8.7 and the 1.9 releases.

TL;DR Bugs like #2133 happen independent of the release cycle, we would easily miss it too even if we do a 'large regression' sweep. 

---------------------

### Regression testing, incremental vs full, Full regression testing is really expensive, if we do good incremental testing then we can avoid full regression testing. Plus, as detailed above, full regression testing is not that useful since we still miss important bugs.

At some point early on in transitioning to Github and Github supported code reviews, we established that PR reviewers would verify testing of features submitted. We've had a number of experiences where we did not test quite thoroughly enough and introducing bugs, sometimes finding out a good bit down the road (I would suggest that even #2133 is an example of this).

So if a reviewer verifies testing, and a PR author writes their tests and does a thorough job in testing, we should be pretty solid on any update on that branch. The alternative is not very pleasant, we do not do such a thorough job testing, introduce bugs, then we get to find them the hard way at some later time.

I think the above adds a lot of pressure and incentive to add more automated tests. If we are doing large regression passes of the entire game by hand once a year, then we have less incentive to automate, (and we have less practice testing and knowing what and how to test, so automation looks all the more difficult). 

Last, to this point, full game regression testing is quite expensive for us. There is simply a lot to cover. If we do a thorough incremental testing job we can avoid the regression testing completely. With confidence in what we merge, and it is backward compatible, there is not much reason to not release it.

Let's say we do hold on to releases though. We could line up a few testers to check things out and then we do the release. This has a few issues that make it not very workable:
- we can and do change the code pretty frequently, the smallest change can break things, testing is invalidated quickly
- testing takes a long time to cover the full game, this testing is quickly invalidated due to the above. Not to mention any communication delays that make the process stretch out longer
- using that process over the last few releases, we still missed stuff, it was not very effective in the end
- holding up the release adds more changes on top, potentially more changes to bugs which we are not going to discover until users start using the software (our testing was long over for #2133 for example, keeping that as a prerelease would have delayed discovery even longer). With slow discover, reverting code is more difficult.

--------------

### Bugs aside, there are a number of problems with slow release cycles 

If release frequency would not have changed the fact that we pushed bug #2133 to 'production', what good is it to release often, what problems does it solve? Turns out we solve quite a number of problems:
- it is easy to revert the last change and restore 'master' to a bug-free state. The 'revert' can be reverted again on a new branch that fixes any original problems. If for examle the bug was added some months ago, the updates will not be fresh and a revert may no longer be clean (we may have fixed variable names, or moved files around in packages, etc...)
- we get a lot of practice releasing. It's a problem that not the whole team knows how to do releases and feels authorized to do so. For an OSS project, it's critical we have no one individual be responsible for any task, that instead, we delegate to a team. Right now it's basically only me that does the releases from Github, it's a problem.
- we fix our testing culture in a number of ways. On the part of branch authors - why bother writing unit tests and testing very thoroughly if a bunch of regression testers are going to come along and find any edge cases that were not taken care of anyway? Reviewers could be thinking the same. 
- we put pressure to automate testing. The more we automate, the less we test manually. Even then as features grow, we would have less and less manual testing to do. Otherwise, without that pressure, as we add features, the scope and cost of regression testing would grow. As that happens we start skipping it, introduces bugs, we start avoiding changes and hacking things in, which is a feedback loop and the problem grows worse over time.
- we never quite developed a pattern and practice of releasing. It's not a simple routine no-thrills event. If we release frequently we'll refine the process, have pressure to automate it (not just the testing, but the actual release process itself), and we'll be in practice as a team.

If we release frequently, we won't have to publicly ask the TripleA community if anyone has sunk the 20 hours into regression testing, if all changes since then were solid and re-tested, and if each OS was verified, etc... It is quite the savings.

-------------

### Frequent releases = good, people want their contributions to go live, a quick turn-around should encourage more contributors

There are some additional benefits to frequent releases. Notably, I think it would encourage contributions. Observing a project and knowing that new changes are released to 'production' in a few days after submitting a PR is very appealing. Compared to some never ending release process that dragged out over months, changes between 1.8 and 1.9 were not released for around 9 months.

-----------------------

### Releasing often is a straight forward process for rolling forward and back and addresses the problems we have seen and are seeing.

***It does seem to me, all of the above is fixed or at least greatly helped by simply releasing often. That means we simply uncheck the 'prerelease' box after merges. If there are problems that we miss, and we will, even if we do regression testing,  then we go back in and check the prerelease boxes until we have a good version as the latest. That then gives us time to fix master, hopefully with an easy revert that can itself be reverted with a fix when one has been identified. Then with master fixed, we uncheck the 'prerelease' box on the latest

------------ 


Sorry for the wall of text, there are a number of problems and nuances to not releasing very frequently, I wanted to avoid the detail, but the conversation moved there. I've outlined a number of problems by having slow releases, or allowing releases to be considered 'unstable'. I'm very much okay with exploring alternatives, but with the above problems identified, I don't think there are any. At the moment we have a good level of automation, we are talking about a solution that involves:  "click the check box more often"

**Comment by prastle** [comment]
A minor input... But what all seem to fail to recall is
When you release a new stable version the lobby mods are who get to deal with it.    
When games are locking and crashing ....
I would vote for a more stable ... stable release

JMHO

Pras

**Comment by DanVanAtta** [comment]
> When you release a new stable version the lobby mods are who get to deal with it.

Indeed. The argument is shorter release cycles without full regression tests (that miss things anyways) would have us to do very thorough testing before we change anything. Then the idea is we change small things very incrementally with good testing and hopefully very few problems.

**Comment by prastle** [comment]
i hear ya
except
one BIG problem you are missing
Most in lobby visit a few times a week or less
so when ya finally convince them to try new one and it explodes in their faces and they waste hrs screwing around it makes them upset.

Thus why i asked for test lobby like old days

**Comment by DanVanAtta** [comment]
Ignoring backward incompatible changes, I think the questions I have for the group at this point are:

How often do we release code updates?
How much and when do we test?

I think we've seen issues if we do not release pretty often and secondly when we try to do test-the-world regression tests. My answers are to test everything at time of merge and shortly after. Then release as often as we can.

**Comment by DanVanAtta** [comment]
> so when ya finally convince them to try new one and it explodes in their faces and they waste hrs screwing 

I probably would not expect this for a few months at a stretch, probably 4-6 months. For the most part we should be shooting for backward compatible changes and not merging anything incompatible. It's an exceptional case when we change major versions and we would hold releases since it would force upgrades.

**Comment by prastle** [comment]
Simple point
When you are happy
with a new release
Add a second download
in forum link or on website
called test lobby so all can test every map l.ike old days

**Comment by prastle** [comment]
atm in the last week 50 guys have reinstalled and uninstalled
and reinstalled
why di we promote this?

**Comment by prastle** [comment]
Or to be specific why did I?

**Comment by DanVanAtta** [comment]
@prastle Check out my latest comment in: https://github.com/triplea-game/triplea/pull/2137, it'll be easy to config a test lobby from the game. Skipping the long story, the old configuration for a test lobby has some subtle problems.

> why did we promote this?

:cry: The last release we did not have that much confidence in, and we needed to get it out there so we could finally see what problems we had. We were good on regression testing for a moment but then merged code. We are updating code too often for us to rely on stop-the-world regression testing and need to do a better job IMO at testing merges. I mentioned for us not to advertise the release for a few days... Which did pass.. This is regrettable. IMO we need to focus on better testing (and more automated testing so we can test more things more often). Again though, this should not be confused with release ferquency, the fact we had such bad bugs is I think because we don't release so often (so we allow these bugs to creep in and then let the users test it for us)

**Comment by prastle** [comment]
To clarify.  I 100% agree that the only way issues will be found is by playtesting.  

But that needs to be done by people that are expecting an error not the guys that just join to veg out.

Completely up to u

Just so realize every guy that gets annoyed is one less player.

Done my spiel

**Comment by prastle** [comment]
so to continue for a second

We need a test lobby which you have done Awesome!

thats is accessible by all... it isnt

end of story

**Comment by prastle** [comment]
The only way you will have a test lobby is if you make a link in downloads on website for the guy that just wants to cliclkk

**Comment by prastle** [comment]
What i did to join (and it was empty) will never fly

also had to login anonymous

**Comment by prastle** [comment]
visited a few times thus it kinda pointless.  We need people testing there and playing

**Comment by DanVanAtta** [comment]
To test any client updates you don't need to connect to the test lobby. Most updates do not need connecting to the test lobby. So you can find players with same version on the old lobby and test.

**Comment by prastle** [comment]
ah man
I give up do what ya wish but people here play a different person everyday
we dont want to check why it locked this time?
daily every hr
jmho
gl

**Comment by prastle** [comment]
to explain better
I shouldnt have to ? what engine he has before i play a name that has been here 6 years
if he hasnt been on in 2 months

**Comment by DanVanAtta** [comment]
> But that needs to be done by people that are expecting an error not the guys that just join to veg out.

This is a good point. It does point to the need to two versions of the game. To get the latest, people can go to: https://github.com/triplea-game/triplea/releases/, perhaps we can link that on the download page, yeah.

So maybe the answer is tracking how we do testing. A thread where people can say how many hours they played on a certain version maybe would be useful. It's a problem how to track testing when code is changing pretty frequently. My only answer is to make sure we don't break anything to begin with, in which case it does not matter that people are downloading the latest. @prastle adervtising a mass upgrade probably should be coordinated first

**Comment by prastle** [comment]
u guys listed it as stable

thus i wanted them all in stable
 SO NEXT time ya list a stable
MAKE SURE ITS STABLE!
;)

**Comment by prastle** [comment]
or dont use the lobby as ur guinea pigs ;)

**Comment by prastle** [comment]
also to explain further lobby mods have to deal with what u list
and explain where to send error messages etc
otherwise you get none


because


until last week all links broken


thus why i was happy when i seen ya had new stable (when they clicked help it went to somewhere they read)

**Comment by prastle** [comment]
most used .3267
 until last week
end of story

**Comment by DanVanAtta** [comment]
Thanks @prastle , you do raise good concerns. WE do need to come up with some technical solutions, and or better process. What we have been doing for testing is too time consuming and we have been missing big things anyways - it is not working. A similar process before only seemed to work before IMHO because things were changing much more slowly in the code.

I believe we're looking at a scaling problem, we need to make a lot of changes in code and because of its structure and state of testing, we're very likely to break things. So we're constantly breaking stuff and then spending forever to find it, it does not work given how much we have to do... So again, something needs changing, IMO simply stop relying on regression testing and move that effort forward. Couple that with more automation, and I hope that would be a better path.

**Comment by prastle** [comment]
Agree 100% dont slow down 

BUT

Throw a test looby out there thats downloadable by all!

**Comment by prastle** [comment]
Yes you will have to update every time you find a new version that you think might be stable.  Except everyone testing it is looking for an error.  The truth is every time a release is out it seems to thin lobby.  Only advise i can give ya. GlHf

Pras

**Comment by DanVanAtta** [comment]
> Throw a test looby out there thats downloadable by all!

It's about to be pretty easy:
1. https://github.com/triplea-game/triplea/releases
2. Add test lobby IP to settings by opening the engine preferences window

Does it need it need to be much simpler? Can it be simpler? Where is most difficult part? The IP address and port may change. AS for peopel not being in a test lobby, that's probably expected most game updates should be backward compatible, so we don't need people in a test lobby, they can play anyone.

**Comment by DanVanAtta** [comment]
> Yes you will have to update every time you find a new version that you think might be stable

That is part of my point, we've no business pushing any code to master that would cause an "unstable". Are we okay to accept bugs? IMO this is exactly using other people as guinea pigs, we should be merging code with lots of confidence and not making it unstable to begin with. Hence, invest in testing at the time of merge, and make it clear that is when we need to make sure our updates our stable. Given the cost of testing, I think that is likely optimal for our time usage.

**Comment by DanVanAtta** [comment]
Given the conversation, I'm seeing two options:
1. Agree to do all testing at time of merge, bump latest version whenever we can
or
2. Come up with a way to track testing so we know when something has been tested well enough to be called a stable.

I'm open to additional options/approaches, ideally we can keep things low tech and on a path where we can keep automating things and reducing our overhead.

**Comment by prastle** [comment]
Prob is i have a hard enough time teaching many how to dl.  I dont truly believe they are that inept.  I believe they just are not interested in having to do so.  Add a button on website for test lobby.  Update it every time ya want something tested.  The Devs and gamers that care will do so.  THEN when a few weeks have gone buy and no complaints we push a stable release.

**Comment by prastle** [comment]
commenting on first working on second and third u type faster thatn me ;)

**Comment by prastle** [comment]
keep dev sep from stable ... when ya reach a stable ur happy with move to test lobby ... if it passes test lobby we bump to stable and promote

**Comment by prastle** [comment]
but just to explain why your having probs 300? maps 30 played

**Comment by prastle** [comment]
so test lobby needs a few of each played
  now thats a headache

**Comment by DanVanAtta** [comment]
why have a test lobby when 95% of the time we can connect to the existing lobby and do all the same testing? We also get some additional compatibility testing as well. 

The test lobby never had a large population, I am skeptical we will have more luck now. Games would have to be scheduled, likely we would change a couple of things in that time before the two players could lock on the same and latest version.

**Comment by prastle** [comment]
BECAUSE
 the testers are using new instead of playing for an hr or 2 and then crash

**Comment by prastle** [comment]
testers know it might crash save every rnd

lobby are all happy till their game FUCKS UP!

**Comment by DanVanAtta** [comment]
To be sure, I'm implying that we get rid of the concept of dev, that we always have a stable. A 'dev' means we allow bugs to be introduced, and once we do that, we don't always find them when we go looking for them. Back in teh day it made sense to do that and spend a month out of six testing. Given we can merge code at much less cost, it really does not make sense to do that anymore.

**Comment by prastle** [comment]
To clarify how happy are you 2 hrs in if game crashes with no save?

**Comment by prastle** [comment]
u keep bringing up cost
 what ya fail to realize is they would all pay some penutsif it worked

**Comment by DanVanAtta** [comment]
A point I'm trying to make is that if we release more often with more discipline, the number of times there will be such a crash will go down. But ultimately those bugs are not directly related to release frequency anyways, and there are other problems that make such bugs more likely when we don't release often. Next, reporting and feedback can be delayed, and getting such problems fixed takes more time with slow releases.

**Comment by prastle** [comment]
think i have proven that

**Comment by prastle** [comment]
i agree it has nothing to do with frequency

it needs play testing thus ya need an unstable lobby

**Comment by DanVanAtta** [comment]
FWIW, when I refer to cost, usually I'm referring to time/labor.

**Comment by prastle** [comment]
ah
kk

**Comment by DanVanAtta** [comment]
You can play test by connecting to either lobby, you only need to download the latest.

**Comment by prastle** [comment]
right
except then lobby guys have to suffer as well

**Comment by prastle** [comment]
i noted 0 in your test lobby

**Comment by prastle** [comment]
so my sug is simple get them to test give them a simple link
 but up to u

**Comment by DanVanAtta** [comment]
yeah, if one person knows they have a test version and the other one does not. That's a problem. A "Testing" room in gitter might be of help here. Start a community/conversation on the topic. We probably need a lot more dedicated chatter on what has been tested and what needs to be.

**Comment by DanVanAtta** [comment]
This is the link @prastle : https://github.com/triplea-game/triplea/releases
We can add that to the downlaod page if it helps.

**Comment by prastle** [comment]
bro i have had it copied for over a year
how do u think they got the updates l;inks were broken till last week ... gigglez

**Comment by prastle** [comment]
anyways get some sleep me i need to long day tomorrow u have heard my thoughts
the one thing i would like to repeat is triplea LIVES because it has live play
if lobby dies so does triplea
thus ya need to keep lobby w2orking not releases that crash the guys that cant use their computer
just my humble  opinion

if i can help in any way let me know gn bro

**Comment by ssoloff** [comment]
@DanVanAtta 

> We are updating code too often for us to rely on stop-the-world regression testing and need to do a better job IMO at testing merges. <...snip...> IMO we need to focus on better testing (and more automated testing so we can test more things more often).

I don't think how we manually test PRs today is sustainable.  I personally get a little burnt out having to do so many manual test iterations, sometimes over the course of several days. 😫  As you said, the way forward is increasing our automated test coverage.

Therefore, I think the dev team needs to start encouraging each other to include more automated tests with each PR, where possible (e.g. lack of tests for UI code at this time is understandable).  I don't want to add a "thou shalt write unit tests with 80% coverage" commandment to the contribution guidelines, but I do think there should be a certain level of expectation that most PRs include a reasonable amount of automated tests to a) ensure a quality product for the users, and b) respect each dev's time and commitment by ensuring what we submit is solid.

I'm not letting myself off the hook here.  While I TDD all new code I write, I've caught myself skipping writing tests when fixing bugs, adding features in existing code, and unsafe manual refactorings.  But a lot of times I submitted a PR anyway when I could have done a little up-front refactoring to make the code more testable beforehand.  In retrospect, I would have welcomed another dev calling me out on those decisions and holding me accountable. :smile:

And I'm not suggesting increasing the automated test coverage with all future PRs will be some sort of panacea.  I think you did a good job explaining a few months ago how not being able to automate compatibility testing at this time is a huge hole in our release process.

Sorry if I came across preachy; I've been up for 20+ hours and am too tired to try and re-write it. :smile:

**Comment by DanVanAtta** [comment]
Lots of agreement, interesting conversation. We have a number of threads and concerns, any follow up or ideas that have been inspired, let's follow up in new independent issues.

To sum up, we do face a few key problems:
- how to increase testing confidence? Particularly when we have a code structure that is hard to unit test
- how to release well? The measure of success would be that we don't have to do follow up bug fix releases after a stable release, and never should we have to outright roll back to the previous stable release.

I've spawned a few new issues to follow up on some topics:
- #2139 : discuss better tracking of testing during releases, essentially this is a continuation of the them of a release length from this issue. I think we will never find all bugs in a PR review nor regression test, so it does seem the answer is to devise a way to converge on releases that are stable. I suggest a spreadsheet in #2139, that might be a good way to go, I'm sure there are some other good ideas to consider

- #2138 : pick up the theme of how we can do testing better. There is a lot of manual testing, we are struggling as the test burden increases. In this discussion I'm hoping we can also look at TripleA testing a bit more broadly, I think designing a strategy for integration testing would be of benefit, and something we can add to give us more confidence we did not break anything
